### DISCUSSION TITLE
Execute Listen and Unlisten while waiting on Notifications

### DISCUSSION THREAD

**Discussion by jdabbs003** [jackc/pgx#2087] [Q&A] [answered]
I need to listen and unlisten while also waiting for notifications. What's the best way to do this -- two goroutines, the first to receive (via WaitForNotification) and dispatch notifications, and the second to maintain subscription lifecyles (via Exec "listen" and "unlisten")? I suspect thats not a good idea, but I don't know how else to do it. Thank you.

**Comment by jackc** [comment, marked as answer]
`*pgx.Conn` is not concurrency safe, so you can't use two goroutines. You would need to cancel the context passed to `WaitForNotification`, then execute your `listen` or `unlisten`, then recall `WaitForNotification`.

**Reply by jdabbs003** [reply]
That makes sense. Are there any race conditions I need to think about, such that a cancel can cause pg notifications to be lost? Looking at the code it doesn't seem like it but it never hurts to ask.

**Reply by jackc** [reply]
Cancel is safe.
